### PR TITLE
[API][Web][SIMS #1185] - Correct Address Blocks - first part of the story

### PR DESCRIPTION
**Pull request by ann-aot**
-  new address format 
{
(optional) "phone": "Phone Number",
"<address/mailing>": {
"city": "Ottawa",
"country": "123",
"provincestate": "Ontario",
"postalCode": "K1W4Y5",
"addressLine1": "123 Ontario Street",
"addressLine2": ""
}
}
- new Primary Contact Blocks
{
"email": "test",
"phone": "test",
"lastName": "test",
"firstName": "test"
}
- unused old code clean up

### CODE REVIEW

File: sources/packages/api/src/esdc-integration/msfaa-integration/msfaa-files/msfaa-file-detail.ts [hunk 1]
```diff
@@ -49,7 +49,7 @@ export class MSFAAFileDetail implements MSFAARequestFileLine {
     record.appendWithEndFiller(this.addressLine1, 40, SPACE_FILLER);
     record.appendWithEndFiller(this.addressLine2, 40, SPACE_FILLER);
     record.appendWithEndFiller(this.city, 25, SPACE_FILLER);
-    record.appendWithEndFiller(this.province, 4, SPACE_FILLER);
+    record.appendWithEndFiller(this.provinceState, 4, SPACE_FILLER);
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/esdc-integration/msfaa-integration/msfaa-files/msfaa-file-detail.ts:52]: If `provinceState` is not going to be populated by countries other than Canada we should add a null check.
```typescript
record.appendWithEndFiller(this.provinceState ?? "", 4, SPACE_FILLER)
```

File: sources/packages/api/src/route-controllers/designation-agreement/models/designation-agreement.model.ts [hunk 2]
```diff
@@ -35,7 +38,7 @@ export interface LocationsDesignationsDto {
   designationLocationId?: number;
   locationId: number;
   locationName: string;
-  locationData: any;
+  locationData: InstitutionLocationInfo;
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/designation-agreement/models/designation-agreement.model.ts:41]: Maybe not part of the PR but maybe would be good to have the `InstitutionLocationInfo` renamed to `InstitutionLocationData` to be aligned with the column name. Just a suggestion.

File: sources/packages/api/src/route-controllers/institution/institution.aest.controller.ts [hunk 3]
```diff
@@ -75,12 +75,16 @@ export class InstitutionAESTController extends BaseController {
...
       operatingName: eachInstitution.operatingName,
       address: {
-        addressLine1: eachInstitution.institutionAddress.addressLine1,
-        addressLine2: eachInstitution.institutionAddress.addressLine2,
-        city: eachInstitution.institutionAddress.city,
-        provinceState: eachInstitution.institutionAddress.provinceState,
-        country: eachInstitution.institutionAddress.country,
-        postalCode: eachInstitution.institutionAddress.postalCode,
+        addressLine1:
+          eachInstitution.institutionAddress.mailingAddress?.addressLine1,
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/institution/institution.aest.controller.ts:79]: I believe that `mailingAddress` will always be present right? Do we need null checks?

File: sources/packages/api/src/route-controllers/student/student.controller.ts [hunk 4]
```diff
@@ -562,21 +561,23 @@ export class StudentController extends BaseController {
...
+    const address = student.contactInfo.address;
     return {
       firstName: student.user.firstName,
       lastName: student.user.lastName,
       email: student.user.email,
       gender: student.gender,
       dateOfBirth: student.birthDate,
       contact: {
+        address: {
+          addressLine1: address?.addressLine1,
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/student/student.controller.ts:573]: The address should not be null at this point.

File: sources/packages/api/src/esdc-integration/e-cert-integration/models/e-cert-integration-model.ts [hunk 5]
```diff
@@ -38,7 +38,7 @@ export interface ECertRecord {
   addressLine1: string;
   addressLine2: string;
   city: string;
-  province: string;
```
Comment (dheepak-aot) [inline sources/packages/api/src/esdc-integration/e-cert-integration/models/e-cert-integration-model.ts:41]: Were we passing the postal code data to province key before?

File: sources/packages/api/src/route-controllers/utils/address-utils.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import { AddressDetailsModel } from "../../services";
+import { AddressDetailsAPIOutDTO } from "../models/common.dto";
+
+/**
+ * Util to transform address details for formIO.
+ * @param addressDetails  is the object that has
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/utils/address-utils.ts:6]: There is an extra space here _addressDetails is_.

File: sources/packages/api/src/route-controllers/utils/address-utils.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+): AddressDetailsAPIOutDTO {
+  return {
+    addressLine1: addressDetails.addressLine1,
+    addressLine2: addressDetails.addressLine2,
+    provinceState: addressDetails.provinceState,
+    country: addressDetails.country,
+    city: addressDetails.city,
+    postalCode: addressDetails.postalCode,
+    canadaPostalCode:
+      addressDetails.selectedCountry !== "other"
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/utils/address-utils.ts:21]: Can we have "other" as a const here?

File: sources/packages/api/src/services/student/student.service.models.ts [hunk 8]
```diff
@@ -1,14 +1,10 @@
+import { AddressDetailsModel } from "../address/address.models";
+
 /**
  * Information that must be provided
  * while creating a new student.
  */
-export interface CreateStudentInfo {
+export interface StudentInfo extends AddressDetailsModel {
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/student/student.service.models.ts:7]: I believe that we should simplify the number of model mapping addresses.
Do they have any difference from the one persisted on DB?

File: sources/packages/api/src/services/supporting-user/supporting-user.service.ts [hunk 9]
```diff
@@ -68,7 +68,6 @@ export class SupportingUserService extends RecordDataModelService<SupportingUser
       queryRunner,
       SUPPORTING_USERS_TRANSACTION_IDLE_TIMEOUT_SECONDS,
     );
-
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/supporting-user/supporting-user.service.ts:71]: This type of change should not be part of the PR, IMO.

File: sources/packages/api/src/types/primaryContact.ts [hunk 10]
```diff
@@ -0,0 +1,6 @@
+export interface PrimaryContact {
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/types/primaryContact.ts:1]: Should we keep `AddressInfo` and `PrimaryContact` at this level or move to the entities level?
I would say that we have been creating the types that are related to entities on the entities level. If there are absolutely no modifications at the service level we can reuse them.

File: sources/packages/web/src/components/common/LocationSummary.vue [hunk 11]
```diff
@@ -129,10 +129,13 @@ export default {
     };
 
     const addressList1 = (item: InstitutionLocationsDetails) => {
+      let addressString = item.data.address.provinceState
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/common/LocationSummary.vue:132]: We have a method to format the address, can we use it here?

File: sources/packages/web/src/services/http/dto/Common.dto.ts [hunk 12]
```diff
@@ -56,7 +12,8 @@ export interface InstitutionPrimaryContactAPIOutDTO {
 
 /**
  * Common DTO for address details.
- */ export interface AddressDetailsFormAPIDTO {
+ */
+export interface AddressDetailsFormAPIDTO {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/services/http/dto/Common.dto.ts:16]: Are we missing In/Out in the name?

File: sources/packages/web/src/services/http/dto/Institution.dto.ts [hunk 13]
```diff
@@ -1,9 +1,9 @@
-import { ClientIdType, DesignationAgreementStatus } from "@/types";
 import {
-  AddressDetailsFormAPIDTO,
   AddressAPIOutDTO,
-  AddressInfoAPIOutDTO,
-} from "@/services/http/dto";
+  ClientIdType,
+  DesignationAgreementStatus,
+} from "@/types";
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/services/http/dto/Institution.dto.ts:5]: Can we move the `AddressAPIOutDTO` and `ContactInformationAPIOutDTO` to the DTO folder?
I belive that anything named DTO should not exists outside the DTO folder.
If needed, we can create a contract we the same schema at the types folder.

File: sources/packages/api/src/database/factories/institution-location.factory.ts [hunk 14]
```diff
@@ -20,7 +20,7 @@ export async function institutionLocationFactory(
     firstName: "FirstName",
     lastName: "LastName",
     email: "Email",
-    phoneNumber: "Phone",
+    phone: "Phone",
   };
```
Comment (henryoforeh-aot) [inline sources/packages/api/src/database/factories/institution-location.factory.ts:24]: Is phone attribute optional ?

File: sources/packages/web/src/types/contracts/Common.ts [hunk 15]
```diff
@@ -13,19 +13,20 @@ export interface Address {
...
   country: string;
   city: string;
   postalCode: string;
+  selectedCountry?: string;
 }
-export interface ContactInformation {
-  addresses: Array<AddressInfo>;
-  phone: string;
+
+export interface InstitutionLocationInfo {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/types/contracts/Common.ts:30]: Please consider moving this to a more specific file, like institution or location.

File: sources/packages/web/src/views/student/financial-aid-application/FullTimeApplication.vue [hunk 16]
```diff
@@ -149,7 +149,7 @@ export default {
 
       const address = studentInfo.contact;
       // TODO: Move formatted address to a common place in Vue app or API.
-      const formattedAddress = `${address.addressLine1} ${address.addressLine2} ${address.city} ${address.provinceState} ${address.postalCode}  ${address.country}`;
+      const formattedAddress = `${address.address.addressLine1} ${address.address.addressLine2} ${address.address.city} ${address.address.provinceState} ${address.address.postalCode}  ${address.address.country}`;
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/student/financial-aid-application/FullTimeApplication.vue:152]: Can we use the address formater here?

**Review by andrewsignori-aot — COMMENTED** [review]
Great an extensive refactor 😉 
Please take a look at the comments.

File: sources/packages/api/src/route-controllers/institution/institution.aest.controller.ts (hunk 3 above)
Comment (guru-aot) [reply sources/packages/api/src/route-controllers/institution/institution.aest.controller.ts:79]: I guess, you are right @andrewsignori-aot in the institution creation formio forms, those are mandatory fields. I think, this was introduced sometimes recently, as I see in Dev, some institutions does not have mailing address. We can safely remove the null checks. 
Also instead of repeating eachInstitution.institutionAddress.mailingAddress, we can create a constant called as mailing address inside the method and use it inside the address block.

File: sources/packages/web/src/components/common/LocationSummary.vue (hunk 11 above)
Comment (guru-aot) [inline sources/packages/web/src/components/common/LocationSummary.vue:132]: Please rename the addressString variable to stateCityPostalCodeString or something. 😄  I know its tough to name these. Just a suggestion.

File: sources/packages/api/src/route-controllers/institution/institution.institutions.controller.ts [hunk 17]
```diff
@@ -161,7 +161,7 @@ export class InstitutionInstitutionsController extends BaseController {
 
     return {
       ...institutionDetail,
-      mailingAddress: transformAddressDetailsForForm(
+      mailingAddress: transformAddressDetailsForAddressBlockForm(
```
Comment (dheepak-aot) [inline sources/packages/api/src/route-controllers/institution/institution.institutions.controller.ts:164]: why are we doing this transformation? The form can be refactored to align with the Model/DTO right?

File: sources/packages/web/src/views/aest/student/Profile.vue [hunk 18]
```diff
@@ -35,11 +35,11 @@ export default {
         email: studentDetail.email,
         dateOfBirth: dateOnlyLongString(studentDetail.dateOfBirth),
         phone: studentDetail.contact.phone,
-        addressLine1: studentDetail.contact.addressLine1,
-        addressLine2: studentDetail.contact.addressLine2,
-        city: studentDetail.contact.city,
-        provinceState: studentDetail.contact.provinceState,
-        postalCode: studentDetail.contact.postalCode,
+        addressLine1: studentDetail.contact.address.addressLine1,
```
Comment (guru-aot) [inline sources/packages/web/src/views/aest/student/Profile.vue:38]: Please create a constant for this studentDetail.contact.address as address and use it here.

File: sources/packages/api/src/route-controllers/institution/institution.institutions.controller.ts (hunk 17 above)
Comment (dheepak-aot) [reply sources/packages/api/src/route-controllers/institution/institution.institutions.controller.ts:164]: May be it is a technical debt.

File: sources/packages/api/src/route-controllers/student/models/student.dto.ts [hunk 19]
```diff
@@ -69,18 +77,17 @@ export interface StudentRestrictionDTO {
 /**
  * DTO object for student details.
  */
-export interface StudentDetailDTO {
+export class StudentDetailDTO {
```
Comment (dheepak-aot) [inline sources/packages/api/src/route-controllers/student/models/student.dto.ts:80]: Just wanted to remind that, if we are changing from interface to class then make sure it is not used in any of the API as payload. If otherwise the payload will not be parsed without using @Allow decorator.

File: sources/packages/api/src/route-controllers/utils/address-utils.ts [hunk 20]
```diff
@@ -0,0 +1,34 @@
+import { AddressDetailsModel } from "../../services";
+import { AddressDetailsAPIOutDTO } from "../models/common.dto";
+
+/**
+ * Util to transform address details for formIO.
+ * @param addressDetails  is the object that has
+ * address details of type AddressDetailsModel.
+ * @returns address info as AddressDetailsAPIOutDTO.
```
Comment (dheepak-aot) [inline sources/packages/api/src/route-controllers/utils/address-utils.ts:8]: As we discussed in the past, we must avoid using object type name for returns.

File: sources/packages/api/src/utilities/address-utils.ts [hunk 21]
```diff
@@ -1,84 +1,23 @@
-import { AddressDetailsAPIOutDTO } from "../route-controllers/models/common.dto";
 import { AddressDetailsModel } from "../services";
 import { AddressInfo } from "../types";
 
 /**
  * Util to transform address details.
- * @param addressDetails is a partial AddressDetailsModel.
+ * @param addressDetails is the object
+ *  that has address details of type
+ *  AddressDetailsModel.
```
Comment (dheepak-aot) [inline sources/packages/api/src/utilities/address-utils.ts:8]: Avoid object type name

File: sources/packages/web/src/types/contracts/AddressContract.ts [hunk 22]
```diff
@@ -0,0 +1,17 @@
+/**
+ * Common DTO for Address.
+ ** This DTO is returned as API response body.
+ */
+export interface AddressAPIOutDTO {
```
Comment (dheepak-aot) [inline sources/packages/web/src/types/contracts/AddressContract.ts:5]: Why is DTO created in contract.ts?

**Comment by dheepak-aot** [comment]
Great job @ann-aot. Articulating the connections of these changes in client side especially form.io  is not easy.

As per our discussion and also what I saw from common address API DTO, It was having class validators in an assumption that dry run submission is going to take care of all the validations. But on update institution profile as I mentioned API gets only a partial DTO and hence it is validated through class validators and not the dry run.  If we decide to proceed with class-validators further, the AddressAPI dtos need to use more specific class validators.

File: sources/packages/api/src/database/factories/institution-location.factory.ts (hunk 14 above)
Comment (ann-aot) [reply sources/packages/api/src/database/factories/institution-location.factory.ts:24]: no its not optional

File: sources/packages/api/src/esdc-integration/e-cert-integration/models/e-cert-integration-model.ts (hunk 5 above)
Comment (ann-aot) [reply sources/packages/api/src/esdc-integration/e-cert-integration/models/e-cert-integration-model.ts:41]: good catch, its `provinceState`

File: sources/packages/api/src/route-controllers/institution/institution.aest.controller.ts (hunk 3 above)
Comment (ann-aot) [reply sources/packages/api/src/route-controllers/institution/institution.aest.controller.ts:79]: as discussed the null check is added for old data,

Comment (guru-aot) [reply sources/packages/api/src/route-controllers/institution/institution.aest.controller.ts:79]: Can we please create a constant for eachInstitution.institutionAddress.mailingAddress, instead of repeating them?

File: sources/packages/web/src/components/common/LocationSummary.vue (hunk 11 above)
Comment (ann-aot) [reply sources/packages/web/src/components/common/LocationSummary.vue:132]: good catch 👍

File: sources/packages/web/src/services/http/dto/Common.dto.ts (hunk 12 above)
Comment (ann-aot) [reply sources/packages/web/src/services/http/dto/Common.dto.ts:16]: comment added, as discussed

File: sources/packages/api/src/route-controllers/student/models/student.dto.ts (hunk 19 above)
Comment (ann-aot) [reply sources/packages/api/src/route-controllers/student/models/student.dto.ts:80]: it's an OUT DTO. i kept it for the API refactoring story, anyway I will rename it `StudentDetailAPIOutDTO`.

File: sources/packages/api/src/route-controllers/institution/institution.institutions.controller.ts (hunk 17 above)
Comment (ann-aot) [reply sources/packages/api/src/route-controllers/institution/institution.institutions.controller.ts:164]: its too keep `selectedCountry`

**Comment by ann-aot** [comment]
> Great job @ann-aot. Articulating the connections of these changes in client side especially form.io is not easy.
> 
> As per our discussion and also what I saw from common address API DTO, It was having class validators in an assumption that dry run submission is going to take care of all the validations. But on update institution profile as I mentioned API gets only a partial DTO and hence it is validated through class validators and not the dry run. If we decide to proceed with class-validators further, the AddressAPI dtos need to use more specific class validators.

dryrun validation is part of my next PR. will discuss it with the team and update it accordingly

**Comment by ann-aot** [comment]
accidentally closed the branch. reopening

**Review by henryoforeh-aot — APPROVED** [review]

File: sources/packages/api/src/route-controllers/utils/address-utils.ts (hunk 20 above)
Comment (ann-aot) [reply sources/packages/api/src/route-controllers/utils/address-utils.ts]: this is a different scenario, please go through the comment if you didn't get it. let me know

File: sources/packages/api/src/utilities/address-utils.ts (hunk 21 above)
Comment (ann-aot) [reply sources/packages/api/src/utilities/address-utils.ts]: comment added above

File: sources/packages/api/src/route-controllers/institution/institution.controller.service.ts [hunk 23]
```diff
@@ -34,6 +34,11 @@ export class InstitutionControllerService {
     }
     const isBCPrivate =
       INSTITUTION_TYPE_BC_PRIVATE === institutionDetail.institutionType.id;
+
+    // {} as AddressInfo is added to prevent old data to break.
+    const mailingAddress =
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/institution/institution.controller.service.ts:39]: Maybe it should be 
```typescript
const mailingAddress = (institutionDetail.institutionAddress.mailingAddress ?? { }) as AddressInfo;
```

File: sources/packages/api/src/route-controllers/utils/address-utils.ts (hunk 20 above)
Comment (dheepak-aot) [reply sources/packages/api/src/route-controllers/utils/address-utils.ts]: What I meant is we must avoid using AddressDetailsAPIOutDTO. for e.g. it can be written as address info as DTO/DTO object instead of using the object name.

File: sources/packages/api/src/utilities/address-utils.ts (hunk 21 above)
Comment (dheepak-aot) [reply sources/packages/api/src/utilities/address-utils.ts]: same reply as above 😊

File: sources/packages/api/src/route-controllers/student/models/student.dto.ts [hunk 24]
```diff
@@ -121,3 +128,17 @@ export class AESTStudentFileDTO extends StudentUploadFileDTO {
 export class StudentFileMetadataDTO {
   applicationNumber?: string;
 }
+
+export class StudentInfo {
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/student/models/student.dto.ts:132]: I believe that at this level the class should be named DTO, right?

File: sources/packages/web/src/composables/useFormioUtils.ts [hunk 25]
```diff
@@ -6,6 +6,7 @@ import { Utils } from "formiojs";
 export function useFormioUtils() {
   // Get a component in a form definition once it is loaded.
   const getComponent = (form: any, componentKey: string): any => {
+    console.log(form.components, componentKey, "++++++++++++++++");
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/composables/useFormioUtils.ts:9]: Please remove the console.log.

File: sources/packages/web/src/types/contracts/DesignationAgreementContract.ts [hunk 26]
```diff
@@ -25,7 +27,7 @@ export interface LocationsDesignationsDto {
   designationLocationId?: number;
   locationId: number;
   locationName: string;
-  locationData: any;
+  locationData: InstitutionLocationInfo;
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/types/contracts/DesignationAgreementContract.ts:30]: Was it renamed on the API, right? Can we rename it on Vue also?

**Review by andrewsignori-aot — COMMENTED** [review]
Thanks for doing the changes. Just minor comments.

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @ann-aot , Thanks for doing the changes.

**Review by dheepak-aot — APPROVED** [review]
Thanks for doing the changes. LGTM 👍

File: sources/packages/api/src/route-controllers/student/models/student.dto.ts (hunk 24 above)
Comment (ann-aot) [reply sources/packages/api/src/route-controllers/student/models/student.dto.ts:132]: I actually coped this interface from a different file and pasted it here. since there is a story to reorganize student API, i didn't renamed the DTO

**Review by andrewsignori-aot — APPROVED** [review]
Looks good 👍
